### ISSUE TITLE
Will the latest CUDA 12.01 work?

### ISSUE THREAD

**Issue by lonnietc** [jan-van-bergen/GPU-Raytracer#38]
Hello again,

I know that the original developer has passed but was not sure if you had picked up any development on this, or were you just posting for Zuen to keep it online.

I really like the quality of the results that I see on the Github site for the GPU-Raytracer and would like to do some work with it towards seeing what it might take to integrated it into the 3D engine that I am working with now, as well as to see what potential and effort it might take to make it really a quality real-time ray tracing engine.

I'm sure that this will be much more work than it sounds above, but the results could be simple amazing and could lay the groundwork for a simpler 3D engine that could actually compete with Unreal Engine and CryEngine to name a few. (Maybe wishful thinking, but you never know.)

Any way I see that a dependency for Windows 10 is CUDA 11 and I was wondering if you had tried to compile it to run with CUDA 12.01 (latest) version?

Thanks again.

**Comment by Pjbomb2** [comment]
I have not picked up development, and am instead working on my own pathtracer(on my github), which I mention because without Zuen, I wouldnt have had the drive to continue it a year ago,
Sorry, but I am not sure if this will work with the latest cuda, havent used it in awhile sorry

**Comment by lonnietc** [comment]
Hello,

I have been able to compile it up with CUDA 11.0.2 and it seems to initially work run by giving me the help screen with:

```
Pathtracer.exe --help
```

I did see some mention of using the Mitsuba examples but not sure how to use them:

```
https://github.com/jan-van-bergen/Mitsuba-Scenes
```

Now the question is how to interactively render the images that you have on the Github site?

I am wondering what a typical CLI invocation looks like.

Any help that you could provide would be appreciated.
Thanks agian.

**Comment by lonnietc** [comment]
Nope, I was mistaken as it will build apparently, but I get a runtime error for both the Release and Debug versions:

```
Loaded BVH 'Data/sponza/models/sponza_344.obj.bvh' from disk
Loaded BVH 'Data/sponza/models/sponza_275.obj.bvh' from disk
Loaded BVH 'Data/sponza/models/Icosphere.obj.bvh' from disk
Loaded BVH 'Data/sponza/models/sponza_381.obj.bvh' from disk
NVRTC output:
nvrtc: error: invalid value for --gpu-architecture (-arch)
```

**Comment by Pjbomb2** [comment]
I am not sure of the error tbh, its been a year since I launched it, sorry

**Comment by desparito** [comment]
Hi @lonnietc , in Device/CUDAContext.cpp, at `CUDAContext::init()`, the best compute capability (`best_compute_capability`) is determined and this variable is passed to NVRTC's --gpu-architecture command line argument as `compute_XX`, where `XX` is this compute capability. However, this might not match with an actual compute (minor) version (you can find the accepted command line arguments in the [user guide](https://docs.nvidia.com/cuda/pdf/NVRTC_User_Guide.pdf), page 17). 

It's likely that the code is passing this incorrect argument to NVRTC, so the best option is to round down this `best_compute_capability` to the nearest major version (e.g. non-existent version 86 becomes existing version 80).
